### ISSUE TITLE
[UX] Add links for quick access to editing things like the layout, theme, etc. of the current page

### ISSUE THREAD

**Issue by laryn** [backdrop/backdrop-issues#2626]
This thought just occurred to me as I glanced over a notification that came in on another thread ("...there does not seem to be a way of seeing exactly which layout is in action for a page, even by inspecting the code." [Source](https://github.com/backdrop/backdrop-issues/issues/2612#issuecomment-292567621))

What about a new little icon in the admin menu that either links to a separate page or has a little dropdown info pane with details about the current page you are viewing?

![home___pr_1782_backdrop_backdrop_testing_site](https://cloud.githubusercontent.com/assets/871421/24821914/0b92455c-1bb7-11e7-86c7-b1a5ab4110c4.jpg)

For example, all in one place, info (and potentially links to edit/adjust) about the:

- Layout used on this page 
- Theme
- Content type
- etc.

That may be overkill but thought I'd throw it out there to see what people thought.

**Comment by Graham-72** [comment]
I think this is a great idea. 👍

**Comment by docwilmot** [comment]
Sound like a great idea for contrib. Pretty sure < 80% would need this <80% of the time.

**Comment by laryn** [comment]
Or maybe an enhancement for [Devel](https://github.com/backdrop-contrib/devel) (additional info on the devel tab)...

**Comment by klonos** [comment]
I know that themes can use theme debug and inspect the page code, but site builders shouldn't have to go through that trouble. We already have contextual menus for editing nodes/blocks/menus/views, so why not improve the contextual menus by adding some more options:

Allow "jumping" straight to the content type of the node being rendered:

![image](https://cloud.githubusercontent.com/assets/2423362/24839714/21e72a50-1da3-11e7-97b9-9fe4f9601f6d.png)

In the same spirit we could add a page-wide contextual menu that "jumps" straight to the layout being used:

![image](https://cloud.githubusercontent.com/assets/2423362/24839846/380800e6-1da5-11e7-8fc2-483f75fd734e.png)

**Comment by laryn** [comment]
@klonos I like the direction you are going but I do still wonder if a better solution is a single contextual gear in the admin menu that gives a link to:
 - Edit node (if it is a node)
 - Manage blocks on current layout
 - etc.

The only reason I say so is to avoid cluttering contextual gears all over the place. 

I find myself often using a block's Contextual menu as though I want to edit that block in particular when really I just want to jump to the layout to manage blocks in general. (So I click to edit the block and then close the modal popup for editing that block, leaving me in the layout block config page).

**Comment by klonos** [comment]
@laryn yes, but...

TL;DR: a) the admin menu is not available to all users that can edit things and b) having the shortcuts of all things that can be edited/configured in the admin bar does not make it clear as to which thing each shortcut is for.

Longer version:

So, contextual.module is enabled by default and so is admin.module. Both of these modules are not required (and can thus be disabled), so placing any shortcut relying on any of these modules entails the risk of loosing that shortcut (and thus its UX benefits). Both of these modules have respective permissions to control who has access to them, but the difference is that by default, out of the box, contextual links require lesser permissions and are thus available to say content editors instead of only the site admins as happens with the admin menu.

Now, I realize that this is not a point to be making because editing blocks and layouts is an admin task, but just wanted to mention this nevertheless. The actual point I want to make is this: disconnection from context (as in what the user is trying to edit/configure)...

An "Edit block" shortcut in the admin bar does not tell the user what they are about to edit. The user either needs to know or guess. Contextual links on the other hand mitigate this problem by surrounding the thing being edited with dotted lines, so the user knows. Now when there are many blocks on a single page, this becomes even harder because you'd need to have more "Edit block xyz" links in the admin bar. That's more or less how context links came to be I think. They placed these "Edit block xyz" links next or close to the things being edited.

Consider the case of the front page with the hero block. If there was a single "Edit page title" shortcut available in the admin bar, how would a novice user know which things is the page title? ...is it the "Welcome to Backdrop" (the hero block title), or the "This is your first post" thing? ...and what about now with page titles becoming proper blocks that can be moved around in different regions or hidden from view? So, point is that having separate shortcuts visually connected to these items makes it clear what you're about to edit. That's why I think context menus are a better fit.

Note: with D7 I found contextual filters to be annoying and it would be one of the 3 modules I'd disable once I finished each installation (the other 2 being overlay and the stock toolbar). In Backdrop I find them very useful though and I think that they should be enhanced.

**Comment by laryn** [comment]
@klonos Certainly I was not suggesting that "Edit block" would go into the admin bar area, just things that were page or site-specific. (e.g. current layout, theme, content type, etc.) I guess it wouldn't have to be in the admin bar and a top-level contextual link could suffice. (I'm not sure if @docwilmot opened a separate issue after this comment, which could relate: https://github.com/backdrop/backdrop-issues/issues/1704#issuecomment-292835327 )

**Comment by stpaultim** [comment]
I opened https://github.com/backdrop/backdrop-issues/issues/3946 a while ago to address this same issue. The ideas here might be a better solution.

**Comment by stpaultim** [comment]
This would solve one of my growing pet peeves. I don't care if it's in contrib or core, but I'd really like something like this. I'd work on a contrib module to get it started, but I'm not sure how. Does anyone want to show me?

I'm likely to bring this up in a future Design/UX meeting and either push for this in core or seek help creating a contrib solution. 

At a minimum, I'd like a quick link to my current layout.

**Comment by philsward** [comment]
My thoughts on this is to redesign how and what the menu does.

Leave The top menu for site related tasks, but add a left menu that handles page specific tasks. I know it's a lot to ask, but copy the essence of how WP handles their UX which I think is why so many people find WP easy and Drupal/BD difficult.

I envision getting rid of tabs like View and Edit and adding them to a left menu, then pulling in information about the page being viewed such as the ID, or Parent menu, etc. Info that site builders would want to know without hitting edit to see it, but also a common set of tasks that is handled by the core theme instead of relying on the site or admin theme.

The info and actions about the current page layout would live in this left menu area, displayed if the user has permission.

I personally would like to see context links greatly reduced because they aren't a very good UX design choice.

**Comment by bugfolder** [comment]
To give people something to look at (and/or throw rocks at), I've put together [PR 3756](https://github.com/backdrop/backdrop/pull/3756) that adds a contextual links menu to the top right of the current layout. Currently, it only contains links to edit the layout (not the theme), but if folks think that we should add more links for the theme, etc. (as in @klonos's [comment above](https://github.com/backdrop/backdrop-issues/issues/2626#issuecomment-292802780)), it should be easy to do so.

There is one issue that I couldn't figure out and would welcome suggestions. In Basis (like on the hope page), the default positioning of the contextual link trigger is hidden by the admin bar. In Seven, it's properly positioned. I added some inline CSS that fixes the Basis problem, but at the cost of in Seven, pushing the trigger downward. (Compare the home page to any admin page in the Tugboat site.) Any frontend/CSS gurus who can suggest a better way to address this issue would be most appreciated!

**Comment by docwilmot** [comment]
First rock: this shouldnt be added to `layout_route_handler()`; its probably best in `layout_preprocess_layout()`.

**Comment by bugfolder** [comment]
I confess I'm puzzled by that. To add the contextual links, we need to wrap HTML stuff _around_ the layout. `layout_preprocess_layout()` creates/alters the variables that get rendered _within_ the layout template, but I don't see how to specify a variable that results in wrapping stuff around the rendered template. Can you amplify?

The docs for `layout_route_handler()` describe it as:

```
 * Route handler callback; Execute the current route item or wrap in a layout.
 *
 * @param $router_item
 *   The menu router item for the page currently being loaded.
 * @return
 *   The fully built HTML content for this page, which will be wrapped in
 *   page.tpl.php.
```

Since it ordinarily returns content wrapped in a layout, it seemed like if a further layer of wrapping was needed, that was the place to do it.

**Comment by bugfolder** [comment]
OK, it looks like I can add a prefix/suffix to the renderer that's in the `$variables` array, so [here's a new PR](https://github.com/backdrop/backdrop/pull/3757) that puts everything in `layout_preprocess_layout()`! (Still has the CSS issue, still solicit suggestions on that one.)

**Comment by olafgrabienski** [comment]
> In Basis (like on the home page), the default positioning of the contextual link trigger is hidden by the admin bar. In Seven, it's properly positioned.

I don't have a solution yet, but wanted to mention the reason for the difference. In contrast to *Seven*, the *Basis* theme puts a padding-top and a negative margin-top to the layout, see https://github.com/backdrop/backdrop/blob/1.x/core/themes/basis/css/layout.css#L26: 

```
.admin-bar body .layout {
  padding-top: 33px;
  margin-top: -33px;
}
```

**Comment by bugfolder** [comment]
Thanks, I just found that, went to write about it and saw you'd beat me to it! That's exactly what caused the problem. Removing those two lines fixes the issue in Basis, and, as far as I can see, Basis renders perfectly well without it, since the CSS in the admin_bar module already makes appropriate room for the admin bar.

Anyhow, I've removed that line from Basis and removed the CSS tweak from the PR, and it looks like the contextual links are working well on both themes (and presumably on any other themes that don't munch the top margin).

I realize that changing CSS for the core Basis theme is not something to be done lightly. I wasn't able to see any deleterious effects from removing those two lines.

**Comment by stpaultim** [comment]
@bugfolder - Thanks. This is great and serves exactly what I've been interested in for a long time. 

In previous discussions, there has been some objection to this idea on usability grounds and the potential to confuse users. I believe that the fear is that users will be unclear about where they are being taken and not understand the implications of what they are changing. I don't quite see this problem, but I'm not ready to reject the concern either.

As someone that uses contextual menus A LOT. I love this. But, I know that some people find contextual menus confusing. I think that a version of this that resolves any usability concerns, would be very popular with site builders/architects and would be great in core. 

But, if there are objections to this. I'd REALLY love to get this into a contrib module. This functionality should be available somewhere.

**Comment by bugfolder** [comment]
On the usability issue, I have really liked contextual menus on Views, teasers, blocks, etc. The idea that I want to change "this thing here" and if I hover over the thing I want to change, there's the gear icon to change it just seems really helpful. And we already have contextual links on all the blocks in a layout (for those with permissions to see them); I don't think one more set appreciably adds to the confusion, but it sure helps the usability, IMHO. So I'd advocate for putting it in core.

That said, the new approach (H/T @docwilmot!) would also be doable in a contrib module, I think.

**Comment by bugfolder** [comment]
> It's the users I build the sites for, who have problems with contextual links.

It might help them for you to disable the Contextual module upon site delivery. Or disable specific permissions, like "administer layouts" (which would take away this contextual menu).

There is a separate issue of whether there's a better admin interface for less-savvy administrative users, and I think that's worthy of discussion and investigation. But as long as we're making use of Contextual module and we have contextual links on Blocks and Views and the like, having a set of contextual links on the Layout seems appropriate.

**Comment by olafgrabienski** [comment]
> Removing those two lines [padding-top and negative margin-top] fixes the issue in Basis, and, as far as I can see, Basis renders perfectly well without it, since the CSS in the admin_bar module already makes appropriate room for the admin bar.

Unfortunately, there *is* a side-effect regarding the footer (cf. [code comment](https://github.com/backdrop/backdrop/blob/1.x/core/themes/basis/css/layout.css#L15)), visible on pages with little content which would be less high than the browser window if the layout wasn't set to a height of 100vh. On such pages, without those two lines, a part of the footer disappears 'below' the window. So, I guess we need a fix without changing the CSS of Basis.

Apart of that, I really like the option to jump directly to the *Manage blocks* page of a layout. (Based on the last comments, I've removed the old `needs - more feedback` label, btw.)

One idea: Can we add the destination to the layout links, like we do for block links? Example: `?destination=node/2` for the *About*  page

**Comment by bugfolder** [comment]
@olafgrabienski, thanks for pointing out the effect of that CSS. (The CSS comment was not clear to me; your explanation made it clear.) I've done some updates to the PR:

* I added the destination to the links.
* To address the positioning issue, I made changes to the Basis CSS that are specific to the circumstances of (a) having the admin bar, (b) having layout contextual links present. So in all other circumstances, there's no change to prior behavior. The change I made of adding padding to the bottom is not ideal, because it means there's always a tiny bit of scrollability to the window. I'm not CSS-savvy enough to figure out a better fix (at least not yet), so I would welcome any further input on possible solutions.
* @klonos had a nice suggestion in [the comment above](https://github.com/backdrop/backdrop-issues/issues/2626#issuecomment-292802780) of showing the name of the layout in the contextual menu, so I tried adding this (which entailed adding a bit of CSS to Contextual module to support "informational" entries in the links list). I kind of like the result, and so would advocate that we keep it; but again, I solicit further input.

![example](https://user-images.githubusercontent.com/29214084/135126627-28c5c696-5c45-4a17-b239-9c9ed79bee1f.jpg)

**Comment by stpaultim** [comment]
@bugfolder - I like the addition of the layout name. I think that does improve the usability of this feature, because it helps the user understand where they are going and what they are changing. It also provides users with more information about where they are and which layout is controlling the page they are on.

I would love for this same information to be available to a user in something other than a contextual menu, but for me this is a huge step forward.

**Comment by bugfolder** [comment]
> I would like to mark this "Works for Me," but that might be premature

Yeah, unfortunately I found a puzzling incompatiblity with the Seven theme. If you visit the page /admin/structure/layouts (or any similar admin page), the existing contextual links trigger (gear icon) for the breadcrumbs block is shifted downward by 20 pixels, so that it vanishes when you try to hover over it (which is really annoying!). This PR doesn't do anything to the existing contextual links, but the problem only manifests when this PR is applied. You can see this on the Tugboat site.

The shift comes from a "margin-top: 20px" that gets applied to the trigger; looks like it might be part of the JS processing (?), since the browser is showing it as applied directly to the element, not via a CSS file. So more investigation is needed.

**Comment by bugfolder** [comment]
...And that is happening because of a bug in the Contextual module.

If one contextual links region encloses another, then Contextual pushes the child contextual links downward by 20px so that the triggers won't overlap each other. However, if the child region happens to be shorter than 20px, then the child trigger gets pushed out its region, so it vanishes when you attempt to hover over it. In Seven, the breadcrumb region is only 19px high, so this problem happens. The vulnerability was lurking there all along; we just didn't have nested contextual links with a short inner region until this issue.

The solution is: in contextual.js, when we perform the shift (by setting margin-top), we clamp the amount of the shift so that the trigger doesn't get shifted out of its enclosing contextual links region if that happens to be short. That could conceivably result in partially overlapping triggers (it doesn't happen here because they're also offset laterally), but even if it does, partial overlap is way less bad than having the trigger vanish.

So I've added this fix to the PR, and both contextual link sets are working nicely on Seven (and elsewhere).

**Comment by jenlampton** [comment]
I'm ambivalent about this living in core. I can see how it would be useful, but I'm comcerned about the impact it will have on themes (and perhaps other layouts) on existing sites. I think an admin-bar tool would be less disruptive than a contextual link for the whole layout.

**Comment by bugfolder** [comment]
Yeah, I understand the potential impact on themes. Because of the extra div wrapper, any CSS selector like `body > .layout` would be broken.

But it sure feels like a missing piece from the current suite of Views, Blocks, Node, etc., contextual links, and given the new-ness of Layouts to folks coming from D7, it seems like having this in front of builders (at least, those who are used to contextual links) seems like a good thing.

**Comment by stpaultim** [comment]
I agree that the lack of something like this is a big oversight. At the same time, I don't want to break existing sites, most of which probably have "contextual menus" enabled, eve if they don't use them (in my opinion). 

The original idea for THIS SPECIFIC issue was to put this information in the admin menu. I think that would be better UX (for most users) and it appears as if it would be less likely to break sites. 

I still like the idea of putting the contextual menu option into a contrib module (unless we can figure out how to make it work in core without breaking existing sites/themes). 

@philsward has suggested an even bigger change (a complete revamp of admin UI) that would also help, but it also a much bigger lift and I think should be treated separately.

I think that moving this to admin menu also makes the idea of adding "theme" and/or "content type" more practical (also very useful).

**Comment by olafgrabienski** [comment]
I'd prefer classical contextual links over something in the admin-bar (but I don't know how to address the CSS issues in a good way). I'm however open to an admin bar approach. As far I can see, we didn't discuss how that solution would look like, did we?

**Comment by bugfolder** [comment]
If we go the admin bar route, how about adding an item at admin/structure/layouts/current-layout that takes one directly to the current-page layout? That wouldn't display the name of the current-page layout, though; you'd have to click through to see it, then click back.

It still feels like if we're going to have and use contextual links in core, then a layout contextual links is appropriate. And it doesn't seem out of place to also put theme-based items into that link list, since both layouts and themes are things that apply to the whole page.

If we do end up with "not in core", I've put together a (potential) contrib module that does the same thing, which you can try out at https://github.com/bugfolder/layout_contextual_links.

One of the things I did in the contrib module was to add an admin setting so that you can choose whether to push down the contextual links clear of the admin bar if the theme tinkers with the top margin (like Basis). (I guess we could do something like this in core, but that would seem a bit specialized and hard to discover.)

I'm also thinking that having the first item in the contextual links list being "informational" (like "Layout: _Title of Layout_") could also be usefully added to contextual links for Blocks, Views, etc. Sure, the dotted outline indicates what the gear applies to, but having the name explicitly would be that much clearer.

**Comment by jenlampton** [comment]
I'd like to see a layout contextual link in core for 2.x, but this sounds
like a great (safe) way to add the feature before then :) thanks for the
contrib module, I'll give it a try!

**Comment by bugfolder** [comment]
Lots of discussion at the 9/30 design meeting about this topic. Decisions coming out of it:

* In general, having contextual links on the layout that tell you what layout you're on are considered good to have.
* There remain concerns about possibly breaking Basis or Basis-like themes or subthemes because of adding HTML wrappers, so it would be desirable to allow users to easily turn this on and off.
* Consideration of alternatives to contextual links (like admin-bar-based solutions) should be moved into a separate issue, with this issue remaining just about the contextual-links solution.
* The bugfix to Contextual module that's currently part of PR https://github.com/backdrop/backdrop/pull/3757 should be moved into a separate issue and PR.
* A near-term solution for people who need this capability can be provided by the Layout Contextual Links contrib module, which will be planned to eventually be moved into core. When it's moved into core, part of the core upgrade process would include deinstallation of the contrib module if it was active.
* There was support for adding some finer-grained controls to Contextual module whereby other modules that provide contextual links could declare what they provide (e.g., via a `hook_*_info()`) and allow turning on/off particular families of contextual links (e.g., on comments).

**Comment by philsward** [comment]
@bugfolder Installed it... Love it.

I haven't had to enable the position option for the Seven theme so yay there but I did for Basis as advertised. Huge win for such a small QoL enhancement.

**Comment by bugfolder** [comment]
I"m working on the last bullet above ("adding finer-grained controls to Contextual module") by adding a Contextual Links configuration page at admin/config/administration/contextual-links. It looks like no new hooks are needed: client modules can add their settings via `hook_form_FORM_ID_alter()` on the settings form.

My question to those interested in this topic is: in addition to the layout and theme contextual links discussed above, what other core contextual links would benefit from configuration settings, e.g., ability to turn the contextual links on/off, adding optional links or information to the existing links, etc.?

I've filed [an initial PR as PoC](https://github.com/backdrop/backdrop/pull/3790).

(BTW, I removed the "contrib candidate" label since the contrib module is now in place, and remaining work is focused on core.)

**Comment by bugfolder** [comment]
On my plain-vanilla backdrop installation, comments don't have contextual links (meaning the gear icon and drop-down menu); they have inline links, like this:
![comment_links](https://user-images.githubusercontent.com/29214084/136675064-dc2a22a9-45df-4c5a-9309-04b1f08ee7e8.jpg)
Is there a setting I'm missing? (The comment module doesn't even contain the (case-sensitive) word "contextual", which I'd expect to find if it had them (e.g., a `#contextual` or `'type' => 'contextual'`).

**Comment by bugfolder** [comment]
@klonos, thx for comment! Support for `backdrop_merged_modules()` now added; also added a test to load the new page.

**Comment by indigoxela** [comment]
I wonder, what/how to test, didn't see anything obvious. Where is that contextual menu for layout (theme?) supposed to appear?

**Comment by indigoxela** [comment]
Hm. It seems to break one of my themes (Snazzy). Not really bad, but it's a bit ugly.

But as this contextual link is usually turned off, it may not be that bad.
At least, we should warn people that this functionality may have negative effects on themes.

**Comment by stpaultim** [comment]
It was my understanding that a concern about how this effected contrib themes was in part why we created a contrib version for now. 

Some of us are pretty concerned about adding anything to core that might break existing sites. The danger in this case might be mitigated by the fact that it will only effect administrative users.

https://github.com/backdrop/backdrop-issues/issues/2626#issuecomment-930511124

It seems to me that one way to get this into core would be to make sure that this feature is disabled on by default on existing sites and possibly enabled on new sites, where it can't break things, because they are new sites. Then anyone with an existing site would have access to the functionality by enabling it and will understand that in enabling it they MIGHT break the theme slightly for admins. 

Maybe there are other ways to make this feature safer. I'm not sure. 

To be clear, I love this feature. Right now, the contrib version of this will be required on ALL my Backdrop sites, so having this in core would be great for me. But, I'm uncomfortable breaking existing sites (but might accept a small break on a few sites for admins only, especially if it is possible to disable or do something to compensate for the problem??).

**Comment by indigoxela** [comment]
> To be clear, I love this feature. Right now, the contrib version of this will be required on ALL my Backdrop sites

Never used it so far. It's a nice little addition, but personally I can live without it.

> ...and enabled by default on new sites, where it can't break things, because they are new sites

It can still break things, for instance if contrib themes look ugly. That it only affects admin users doesn't help much, because the site builders usually are admins and they install themes.

So, I'd suggest to add a warning to the configuration page, that turning this on can have side effects on themes. And I'd also suggest to have this turned off by default - on existing and new sites.

**Comment by bugfolder** [comment]
> So, I'd suggest to add a warning to the configuration page, that turning this on can have side effects on themes. 

Added. I also un-collapsed the positional settings box, to give it higher visibility (since this shift is potentially one of those side effects).

> And I'd also suggest to have this turned off by default - on existing and new sites.

Now turned off by default in the config settings file (for new installs) and in the hook_update_N() hook (for existing sites).

**Comment by klonos** [comment]
> It's a nice little addition, but personally I can live without it.

@indigoxela all of us "chronic" (as our doc says) Backdrop users have been using the "workaround" of the "configure block" contextual link, followed by closing the dialog, but that's a horrible UX workflow from a new user point of view. I disagree re it being a "nice little addition" - it is a UX bug fix.

> And I'd also suggest to have this turned off by default - on existing and new sites.

Yeah, I disagree, but I understand 😓 ...I would like us to figure things out, but looking at the usage stats of the various contributed themes, there are just so many people using them out there - we can't go breaking their sites.

**Comment by stpaultim** [comment]
I am a big fan of getting this into core. I was hoping that the decision to add an on-off switch and have it off by default, might be all that was necessary to get this in. BUT, I just reviewed it and I think it will still need some work and reviewing the discussion, I'm not clear that we have enough folks on board with this idea yet.

1) The admin setting is "contextual links" which promises more than it is. This feature is about a very specific contextual link. Maybe it could be "Layout contextual link" or something more specific. And maybe we could edit the help text a bit (less of it).  

![image](https://user-images.githubusercontent.com/3144571/144537509-e0d14b4c-059f-44da-98ea-21a50cb89b29.png)

2) In the end were are introducing four checkboxes for this setting. I would prefer it if we could make it simpler, but I understand the various choices involved and why so many options. 

3) I'm not sure why I can enable the layout contextual link without the theme contextual link, but not the reverse. Although, I do agree with the order of priority. 

4) I disabled the contextual links module and this happened:

![image](https://user-images.githubusercontent.com/3144571/144537202-9d506258-81b2-45c3-a299-e2a00a13bb24.png)

**Comment by klonos** [comment]
> The admin setting is "contextual links" which promises more than it is. This feature is about a very specific contextual link. Maybe it could be "Layout contextual link" or something more specific

I actually like it the way it is currently implemented. My understanding is that this page will serve as the home for any module that introduces contextual links that require settings for some reason. The fact that we have grouped all layout-specific contextual link settings within a fieldset in that form is enough for me ...other modules can add their own fieldsets or "free-standing" contextual link settings.

> I disabled the contextual links module and this happened

Good catch @stpaultim 👍🏼

**Comment by BWPanda** [comment]
A few thoughts:
1. I like the **idea** here! I think a way of going straight to the configuration page for the current layout/theme is great.
1. I don't like this **solution**. It seems hacky, especially if you need a settings page to disable it (and even have it disabled by default) because it may not work, cause issues, etc. And all the checkboxes on this page seem overly complex and unnecessary for a seemingly simple feature...
1. *If* we keep the current solution, it could use some more styling, as the heading's look a bit off (colour, padding, etc.):
![image](https://user-images.githubusercontent.com/2385329/144544765-0427cada-76ce-4915-b8be-e5f1f501396a.png)

I'm sorry I haven't read the whole thread here (too long), so this has likely been discussed already, but as an alternative solution (and IMO, a better one) what's wrong with putting these links into the Admin Bar? I think this is what @laryn was originally suggesting in the OP, so why is that no longer an option?

My thought process is that contextual links are *contextual*, meaning they're put in a certain place to edit that certain part of the page (blocks, nodes, etc.). If you scroll down past a block, you can't see the contextual links for that block anymore because you're no longer in the context of that block. However these layout/theme links are not *contextual*, they're for the page as a whole. So I feel contextual links aren't the answer here, and hence my question about putting them in the admin bar.

**Comment by stpaultim** [comment]
> but as an alternative solution (and IMO, a better one) what's wrong with putting these links into the Admin Bar? I think this is what @laryn was originally suggesting in the OP, so why is that no longer an option?

In my opinion, putting it in the admin bar is still an option. However, I think it's in the contextual links area because that's the preference of @bugfolder and he wrote the PR. ;-) 

As I review the issue, I find that opinions on this issue are quite divided. I found two people (now three) with a clear preference for the admin bar. Three people that clearly prefer contextual links. One person expressed an interest for a new solution in a left hand side bar. 

Personally, I'm for any of the above (although I suppose if forced to choose, I think it would be easier to discover in the admin bar). 

@klonon wrote this detailed explanation on his preference for contextual links: https://github.com/backdrop/backdrop-issues/issues/2626#issuecomment-294099898

**Comment by BWPanda** [comment]
Thanks @stpaultim. I just read through the comments up until this year (more reading to go), but wanted to note that @klonos' advocation for contextual links seemed to be because he thought @laryn was suggesting moving *all* contextual links to the admin bar. Since that's not the case, I wonder what his thoughts are now? Maybe I'll find out once I keep reading... 😉

**Comment by BWPanda** [comment]
I like @philsward's idea about a sidebar/menu. Use the admin bar for site-wide tasks/navigation, and then a sidebar/menu for page-specific tasks/info. Doesn't Drupal 8+ have something like this for node edit forms?

In any case, I think this is a larger, separate issue. Once we implement a solution here for the layout links, we can later move them to the new sidebar along with the other links Phil mentioned.

**Comment by olafgrabienski** [comment]
@BWPanda Re the styling of contextual links, there's a related issue: https://github.com/backdrop/backdrop-issues/issues/5274

Regarding the placement (admin bar vs. contextual links) I'm open for both suggestions, but I couldn't find a realistic mockup of the admin bar idea.

**Comment by bugfolder** [comment]
@stpaultim wrote:

> The admin setting is "contextual links" which promises more than it is. This feature is about a very specific contextual link. Maybe it could be "Layout contextual link" or something more specific. 

As @klonos mentioned, the reason for making this config page "Contextual links" was to create a home for any future settings related to contextual links, with other modules adding via `hook_form_alter()` and each group in its own fieldset (like this one).

> And maybe we could edit the help text a bit (less of it).

I'm open to specific suggestions for what to take out. (Note that it was actually expanded somewhat in the last round at the request of another commenter.) Since this page is likely to only be visited once, at initial config, I think we could err on the side of being more informative.

> In the end were are introducing four checkboxes for this setting. I would prefer it if we could make it simpler, but I understand the various choices involved and why so many options.

I share your preference, but don't see a way of reducing things.

> I'm not sure why I can enable the layout contextual link without the theme contextual link, but not the reverse. Although, I do agree with the order of priority.

If you enable the theme information, but don't enable the layout contextual link, there is no place for the theme information to reside. The theme information is part of the layout contextual link. If there's better wording that makes that clearer, please suggest!

> I disabled the contextual links module and this happened:

Yes, good catch. I'll fix that. [D'oh!]

Thanks, @olafgrabienski, for pointing out the styling PR. Yes, the "information" lines in the link will look better when  https://github.com/backdrop/backdrop-issues/issues/5274 is in place.

Regarding the broader issue of "contextual link versus admin bar?" The content of this dropdown changes from page to page (i.e., it is contextual). That seems like a better fit for the widely-used model of contextual links, which show you actions _specific to this page_. I don't see a good fit for putting it into the admin bar (but that could just be me). Are there examples of existing action links in the admin bar that change from page to page?

Also, with the introduction of information styling into contextual links (again, in https://github.com/backdrop/backdrop-issues/issues/5274), we can mix information and actions into contextual links. Currently admin bar items are all action links (I think?), so if we wanted to move all information from this solution into the admin bar, we'd be creating a new use paradigm.

@BWPanda wrote:

> However these layout/theme links are not contextual, they're for the page as a whole. 

But they vary from page to page, so they _are_ contextual. The context just happens to be the entire page, but that's not something that is the same site-wide.

**Comment by bugfolder** [comment]
> I disabled the contextual links module and this happened:...

Now fixed.

**Comment by indigoxela** [comment]
Took a closer look today and figured out a fix for Snazzy.

Per default the contextual link regions are set to "position: relative", which causes a shift of the top region in Snazzy (and potentially has effects on other themes, too).

The "position: relative" is usually needed to let the .contextual-links-wrapper inside that region to behave "relative" (scroll with the region). But that's not needed for the layout, because the layout always starts on top of page - correct me if I'm wrong.

Setting .contextual-links-region-layout (and only that) to position static, would minimize the side effects on themes, I guess, because then the wrapper inside isn't considered as part of the page/layout (width, height...). Might need a little more experiments, though.

I'm suggesting that to prevent this issue to stall on small problems, that would be easy to fix.

**Comment by bugfolder** [comment]
I squashed the PR and rebased it on 1.20.3.

I then went through all of the themes on b.org that had at least 40 installs (roughly half of them) and tried this out on them, looking at the home page and user/1 page (with admin bar). Here's what I found. ("Needs shift" means that one needed to check the "Shift for...theme" box in the settings.)

* Afterlight: works, but gear has poor contrast against gray background (but that's the case with all contextual links, not just this one).

* Basis: works, needs shift.

* Bartik: works, needs shift.

* Bootstrap Lite: works as-is. Note that this theme puts a header _above_ the layout div, so it's an example where the layout div is not the first item on the page.

* Borg: works as-is.

* Cleanish: works as-is.

* Colihaut: works as-is.

* Corporate KISS: works as-is.

* Lateral: works, needs shift.

* Lumi: works as-is.

* Materialize: works as-is.

* Mero: works, needs shift.

* Pelerine: works as-is.

* Seven: works as-is.

* Shasetsu: works, needs shift. (Current version has some overlapping text with or without this change.)

* Snazzy: works as-is (at least, on the two pages I examined). @indigoxela, can you describe further the problems you experienced?

* Tatsu: works, needs shift, but gear can be partially covered when there is an admin bar alert because this theme has a taller admin bar when there is an admin bar alert. I think that this is because Tatsu theme adds bottom margin to admin bar alerts (but shouldn't).

**Comment by BWPanda** [comment]
In the interest of comparing different solutions for the same problem, here's a PR that adds these links to the admin bar: https://github.com/backdrop/backdrop/pull/3851

![Peek 2021-12-05 22-59](https://user-images.githubusercontent.com/2385329/144745650-6610b637-dd79-42c4-978d-cd975d1ebc2a.gif)

Just like other links in the admin bar, clicking on the main link takes you to the main page for that thing (e.g. manage blocks page for layouts and settings page for themes) with other links (e.g. configure layout) appearing as sub-menu items.

**Comment by indigoxela** [comment]
@bugfolder thank you so much to do the "legwork" to try out all those contrib themes. That should give us a little more confidence.

The problem with Snazzy is that it breaks the split header. The height of the hero inside to gets calculated wrong. This only affects the start page and can get fixed in the theme as well.

TBH I actually like @BWPanda's suggestion for an admin theme link. It makes the functionality easier to discover - but only if the admin menu is turned on. Of course, I'm aware that with the current PR approach, the contextual module is required. (Hopefully my comment doesn't derail this issue. :wink: )

**Comment by BWPanda** [comment]
To summarise this issue, my understanding of it anyway:
- We want an easy way to see and edit the current layout and theme of the page.
- One PR implements this as contextual links. It relies on the Contextual Links module, which is enabled by default. There are some styling incompatibilities with certain themes.
- Another PR implements this as links in the admin bar. It relies on the Admin Bar module, which is enabled by default.
- An idea was proposed to create a UI for turning on/off certain contextual links.

My suggestions for moving forward:
- The contextual links UI form proposal/PR be moved to a separate issue. It's a whole other discussion that deserves its own space apart from adding these specific page links.
- We get feedback here on which (if any) solution we want to see added to core. We can tweak/refine them as we go.

**Comment by philsward** [comment]
> We get feedback here on which (if any) solution we want to see added to core.

+1 admin menu

**Comment by olafgrabienski** [comment]
> In the interest of comparing different solutions for the same problem, here's a PR that adds these links to the admin bar

Many thanks, it's nice being able to compare the approaches in the sandbox websites. I'm still on the fence, though. First thoughts:

- I like the direct visibility of the Admin bar links. Drawback: it needs some space in the bar (only an issue for smaller screens).
- A big advantage of the Contextual link approach is the inclusion of the destination in the links, as in `admin/structure/layouts/manage/home?destination=home`. (IIRC, we don't use this pattern in the Admin bar, so can't or shouldn't add it here.)

Apart of that, there is currently a styling issue with the Admin bar PR when using Basis (or Bartik), see screenshot below.

![backdrop-pr-3851-basis-styling-issue](https://user-images.githubusercontent.com/14188307/144822520-082ff07b-83b9-40c5-88e4-16f5ced6fcbd.png)

**Comment by BWPanda** [comment]
> I like the direct visibility of the Admin bar links. Drawback: it needs some space in the bar (only an issue for smaller screens).

I considered the idea of using just an icon, but I think the text is useful to know what it does. Happy to try either way though.

> A big advantage of the Contextual link approach is the inclusion of the destination in the links... (IIRC, Backdrop doesn't use this pattern in the Admin bar, so we shouldn't add it in this case.)

I'm not so sure... I didn't think to add the destination, but it'd be easy enough to do here. As for if we should, there's an argument to be made for doing so because these links, unlike others in the admin bar, are dynamic and different from the other static links... Again, happy wither way.

> there is currently a styling issue with the Admin bar PR when using Basis (or Bartik)

I noticed this, but it is (in my opinion) a bug in those themes. They style anything with the `.layout` class as 100% of the window height:

![image](https://user-images.githubusercontent.com/2385329/144838799-e004bb77-c4f1-4221-bd57-2790b34f416f.png)

I believe they should be more specific with their styling...

**Comment by indigoxela** [comment]
> I noticed this, but it is (in my opinion) a bug in those themes.

I noticed that problem too, and IMO this needs to get fixed in this PR. The link needs a more specific class, not the pre-existing layout styles, which act on full layouts (hence the class "layout" is correct).

**Comment by BWPanda** [comment]
Fair enough. PR updated to use less generic array keys/classes.

**Comment by indigoxela** [comment]
So we now have two great options. We only need to evaluate the pros and cons ... and decide. :smiley:

**Comment by laryn** [comment]
I lean admin bar over contextuals. However as I just noted in the PR that caching seems pretty heavy in the admin bar which is showing incorrect info when you browse around on the site. 

Also, for text, perhaps "Page info" is slightly clearer than "This page" (and not any longer).

**Comment by bugfolder** [comment]
> Also, for text, perhaps "Page info" is slightly clearer than "This page" (and not any longer).

The setting at admin/config/administration/admin-bar is called "Page links". Would be nice to use the same wording in both places.

For reasons described earlier (and preferring to have less stuff in the admin bar), I prefer contextual links, but in the interest of fairness, I point out that if we end up going with the admin bar, us contextual-link-ophiles can still use [the contrib module](https://backdropcms.org/project/layout_contextual_links) and then uncheck the admin bar item.

**Comment by indigoxela** [comment]
> I just noted in the PR that caching seems pretty heavy in the admin bar which is showing incorrect info when you browse around on the site.

Garbage... that's right, confirmed. Does this make the admin-bar approach impossible?

**Comment by BWPanda** [comment]
Not at all! Working on an update to fix that currently. It'll use the same system as the number of logged in users in the admin bar - that's updated dynamically and not cached.

**Comment by olafgrabienski** [comment]
>> A big advantage of the Contextual link approach is the inclusion of the destination in the links... (IIRC, Backdrop doesn't use this pattern in the Admin bar, so we shouldn't add it in this case.)

> I'm not so sure... I didn't think to add the destination, but it'd be easy enough to do here. As for if we should, there's an argument to be made for doing so because these links, unlike others in the admin bar, are dynamic and different from the other static links...

Thanks for your feedback, sounds reasonable to me! What do others think about adding the destination to these admin bar links?

**Comment by BWPanda** [comment]
PR updated to fix caching issue: https://github.com/backdrop/backdrop/pull/3851

**Comment by indigoxela** [comment]
So we still have two great options here.

I tried to find my personal favorite in a direct comparison (sandbox 1 vs. sandbox 2)... Both solutions have their pros and cons.

The admin menu thing is way easier to discover, but does that link really belong there?
The contextual link setting is hard to find (has to be turned on and I had to lookup the PR code because I forgot again, where).
The contextual approach conforms to how these things are done in Backdrop, but that tiny gear in the top-right corner, potentially covered by the admin-bar, doesn't make it so easy to use.
The contextual link approach might have side effects on existing themes, although they should be fairly minimal.
Both rely on a specific module to be turned on - commonly used ones, of course.
The admin menu approach is zero-conf and works OOTB.

I'm fine with both solutions. With a _slight_ tendency to the admin bar approach.

**Comment by bugfolder** [comment]
Some testing confirms the caching issue for https://github.com/backdrop/backdrop/pull/3851 appears to be fixed.

Right now the admin bar title is "This page" and its configuration checkbox for it is "Page links". Would it be clearer on the configuration page if the checkbox title was also "This page"?

And this is really teeny-tiny, but noticing that the circle-i icon is used for both Reports and This Page, wondering if there's a different icon that would still convey "this page"?

**Comment by BWPanda** [comment]
@bugfolder The names of the different components are generally different than what's displayed in the actual menu:

![Screenshot_20211208-060704_Firefox](https://user-images.githubusercontent.com/2385329/145092982-57cc0991-8346-4007-b35b-2d7c65b0b058.png)

As for the icon, I also considered a cog, but that's also used elsewhere in the admin bar... Open to other suggestions.

**Comment by bugfolder** [comment]
> As for the icon, I also considered a cog, but that's also used elsewhere in the admin bar... Open to other suggestions.

Maybe something like the FontAwesome [file-alt](https://fontawesome.com/v5.15/icons/file-alt?style=solid) that looks like a page?

![file-alt](https://user-images.githubusercontent.com/29214084/145101528-7600753d-90af-473a-bf32-b683a7dd0d75.jpg)

**Comment by stpaultim** [comment]
@bugfolder - Sorry about this, but I have to say that having tried both options - I find the option in the admin bar to be easier to find and use and I suspect that more people will find and use this feature if it's in the admin bar. 

Personally, I'm very comfortable with contextual links and would be happy with either solution, but I think that some users do find them difficult to use and forget about them. 

@BWPanda - Are you aware that the "More Tasks" link is not working correctly on narrower screens with your PR?

**Comment by bugfolder** [comment]
Yup. I've got my preference, but it's not a hill to die on 😉.

**Comment by olafgrabienski** [comment]
I like both the Admin bar and the contextual links approach.

Regarding discoverability of the contextual links, I was asking myself if these contextual links are also hard to find because we're used to look at contextual links at different places: directly at a node, or at a block. Now when I think about it, I'd say the link to the current layout would also fit perfectly at the places where the contextual block links are. (Unfortunately that's not true for the link to the current theme ... but the link to the layout is however more important, in my opinion.)

Conclusion: Otherwise I use contextual links a lot but I'm not sure if I get used to use the new ones :-)

Regarding the Admin bar approach, I'm still interested what others think about adding the destination to the links.

**Comment by indigoxela** [comment]
> Regarding the Admin bar approach, I'm still interested what others think about adding the destination to the links.

I also tried to figure that out for myself - but I'm not sure.

Sometimes it would be nice. A typical use-case: "oops, these two blocks need to get reordered", jump there, reorder, save, ... back to where you started, seeing the desired order.

Sometimes it won't be so jump-there-do-done. If you need to add several things, reorder, fix other stuff, adapt other settings. If this happens in several steps, a redirect after saving is rather obstructive.

BTW, personally I do like contextual links and I use them regularly. The tricky thing in this special case here is, that it's squeezed into the top right corner, possibly even covered by the admin bar in some themes (which requires the setting). For me that makes it hard to discover and use.

**Comment by BWPanda** [comment]
PR updated to use new page icon and add destination parameter to the links. I can't figure out the issue with the "More tasks" link on small screens though...

**Comment by olafgrabienski** [comment]
Thanks for the updated PR! I've tested the PR sandbox, and I think it's good to have the link destination, but I'd also be fine without it.

The "More tasks" behavior on a small screen (via Chrome dev tools) was strange, see the screenshots below:

**(1) After logging in PR sandbox site** (broken)

![backdrop-pr3851-more-tasks-1](https://user-images.githubusercontent.com/14188307/145441122-ed471aef-8516-45a5-b6ac-44cbc999e45a.png)

**(2) After disabling the Page info** (ok)

![backdrop-pr3851-more-tasks-2](https://user-images.githubusercontent.com/14188307/145441143-b5ceab96-0e52-4745-a990-7ae27ec54503.png)

**(3) After re-enabling the Page info** (icon missing)

![backdrop-pr3851-more-tasks-3](https://user-images.githubusercontent.com/14188307/145441148-3575ba49-2ec5-4504-b01d-583a1d94447f.png)

**Comment by klonos** [comment]
> So we now have two great options. We only need to evaluate the pros and cons ... and decide. :smiley: 

Seems like so 🙂 

> ...what's wrong with putting these links into the Admin Bar? ... If you scroll down past a block, you can't see the contextual links for that block anymore because you're no longer in the context of that block. However these layout/theme links are not contextual, they're for the page as a whole. So I feel contextual links aren't the answer here, and hence my question about putting them in the admin bar.

Huh! ...I haven't thought about it that way. I see the point, although my first thought is that it's a good thing that we've recently made the admin bar sticky at the top of the page by default, otherwise these links would still get out of viewport (they still will if people have set their admin bar to not be sticky).

I personally really like the contextual links solution (if only we could figure out a way to have it always "float" on the top-left (top-right in RTL I guess) without affecting themes, then it would be closer to ideal. But:
- it causes issues with some (many?) themes 👎🏼 👎🏼 👎🏼 
- it introduces heaps of configuration options in core, which is the opposite of what we're usually after  👎🏼 
- it needs to be introduced turned-off by default (otherwise we risk breaking themes), which defeats the purpose of adding useful features OOTB 👎🏼 
On the pro's:
- it is an existing UI paradigm, so users don't have to re-train their brains and muscle memory.
- consistency with other contextual links
- with the addition of the "informational", non-clickable items, the actual actions are 2 clicks away (which is one of the UX improvements that contextual links have been created for)

The admin bar solution feels "unnatural" to me, and as I mentioned can have the same undesirable side-effect of getting out of view port if the admin bar is not set to be sticky at the top of the page. But:
- doesn't break themes 👍🏼 👍🏼 👍🏼 
- we can have it enabled by default OOTB 👍🏼 
- doesn't introduce "weird" and complex settings (K.I.S.S. - don't make me think) 👍🏼 
On the con's:
- introduces an inconsistent set of non-contextual "contextual" links in the admin bar 👎🏼 
- we have to figure out a fitting name for the top-level menu item ("This page", "Page links" all seem kinda nonsensical to me) 👎🏼 
- we have to figure out an intuitive icon for the menu item - although figuring out the right label for it may make this a non-issue ...and generally not a biggie.
- the actual actions are 3 clicks/hovers away 👎🏼 
- on narrow screen sizes, the actions get further buried under the "More tasks" item (requires additional clicks) 👎🏼 

Both have pros/cons, but I think I'm leaning towards the admin bar solution, mainly because of the non breaking of themes.

**Comment by docwilmot** [comment]
@BWPanda perhaps you could start a separate issue for your approach so we can separately tag them? I was going to do review but cant say which approach I'm RTBCing if I do!

**Comment by stpaultim** [comment]
As I've thought about this some more and experimented a bit. I like having this in the sticky admin bar when I've scrolled down a long page, scrolling all the way back up to get to the contextual filter is a nuisance.

**Comment by bugfolder** [comment]
> @BWPanda perhaps you could start a separate issue for your approach so we can separately tag them?

It looks like the majority is lining up behind the admin bar solution. I'm fine to close the contextual links option, then you can RTBC the admin bar PR https://github.com/backdrop/backdrop/pull/3851.

**Comment by BWPanda** [comment]
> perhaps you could start a separate issue for your approach so we can separately tag them?

I disagree. This issue was originally opened by @laryn with a suggestion/request for these links to be added to the admin bar. My PR does just that, so it's totally relevent to this issue.

I'm not even sure @bugfolder's alternative approach should be moved to a separate issue, since it solves the same problem but in a different way. We need to decide which approach to go forward with, then test/review that one.

Separate issues potentially mean both features get tested/reviewed/merged, and we don't need two different solutions in core for the same thing.

Having said all that, what I do believe should be moved to a separate issue (as I mentioned [above](https://github.com/backdrop/backdrop-issues/issues/2626#issuecomment-986530559)) is the PR that adds a settings page/form for contextual links. That has ramifications outside of this specific issue, and can be merged regardless of the solution we go with here.

**Comment by bugfolder** [comment]
> what I do believe should be moved to a separate issue...is the PR that adds a settings page/form for contextual links

I've created the separate issue https://github.com/backdrop/backdrop-issues/issues/5404, linking it back to the PR that creates the settings page. Thought I'd hold off on creating a PR for that, though, until we had a specific use case that would put something into that new page.

**Comment by stpaultim** [comment]
Was hoping to mark this "Works for me" - but, I think we're currently blocked by the "more tasks" issue. See: https://github.com/backdrop/backdrop-issues/issues/2626#issuecomment-990039950

**Comment by BWPanda** [comment]
Well that turned out to be an easier fix than I thought! PR updated to fix remaining bug of 'More tasks' menu on small screens.

There's another issue (the sub-menu items in the 'More tasks' menu appear off-screen), but that's an existing issue with the Admin Bar: https://github.com/backdrop/backdrop-issues/issues/4549 So I don't think that should hold this up.

**Comment by stpaultim** [comment]
Works for me - with exception of the sub-menu on "more tasks" which I can't see or access. BUT, I agree that this is another pre-existing issue that should get more attention now. 

I tested it with regular themes and regular layouts. I created a new layout and changed theme to Bartik. None of that cause any problems.

**Comment by klonos** [comment]
@BWPanda there are conflicts in the PR now, so your branch needs a rebase.

My only feedback currently is that I would prefer it if the "This page" menu item was separate from the main menu section in the Admin Bar and the "More tasks" section (its current "grouping" with the "More tasks" doesn't seem appropriate to me). If it was placed in the middle of the admin bar, equidistantly from the other two sections, it would allow it to be visible in narrower screens, without having to expand the "More tasks" menu.

![image](https://user-images.githubusercontent.com/2423362/148674342-62f5391b-5dd3-4991-8e75-9ce5e93f6382.png)
![image](https://user-images.githubusercontent.com/2423362/148674376-50e9509d-dda3-474c-bafe-2aab9b43a8b3.png)
![image](https://user-images.githubusercontent.com/2423362/148674413-bdf98321-e712-44f6-ab7a-ec0fe375da1d.png)

A benefit of making that change would be that we'd work around the "off screen" issue. The rest of the tasks are not as frequently used it seems, whereas I have a feeling that this one will, so #4549 will be more "prominent".

This can be a follow up though.

**Comment by BWPanda** [comment]
PR updated.

I disagree with your suggestion @klonos - to me it looks weird and out-of-place. But file a separate, follow-up issue and it can be discussed there. If others agree, have at it.

**Comment by BWPanda** [comment]
@stpaultim Why the 1.20.4 milestone? This was slated for the next minor release wasn't it (e.g. 1.22, since 1.21's in feature freeze)?

**Comment by klonos** [comment]
@BWPanda the thing I dislike with the current approach vs if it was implemented as a contextual link is that accessing the "configure" link takes more clocks, which defeats the UX improvement.

Before/current:
1. hover over the contextual link
2. click the cog icon
3. click "configure block"
4. close the edit block dialog
(same behavior and amount of actions/clicks on narrow screens)

Initial suggestion:
1. hover over the contextual link
2. click the cog icon
3. click "configure layout" -> 1 less action = UX+
(same behavior and amount of actions/clicks on narrow screens)

Current implementation:
1. hover over the admin bar "this page" item
2. hover over the layout name item
3. click "configure" -> 1 less action = UX+

On narrow screens though, the "this page" item is hidden away, so there's +1 click required -> we fix the UX weirdness of having to close the block configuration dialog, but we do not reduce the amount of actions/clicks.

> But file a separate, follow-up issue...

Yup, will do 👍🏼 ...I would also prefer it if we had a single "Configure [layout_name] layout" link + another "Manage blocks for [layout_name] layout" instead of nested menus (mimic the initial proposal for contextual links). That would reduce the amount of actions further.

> Why the 1.20.4 milestone?

If this was implemented as a contextual link (initial discussions in this issue), it could be considered a task; but the way it is being implemented now feels more like a new feature. So yes, I agree with you @BWPanda that it should be added in 1.22.

**Comment by stpaultim** [comment]
> Why the 1.20.4 milestone? This was slated for the next minor release wasn't it (e.g. 1.22, since 1.21's in feature freeze)?

You are correct. My mistake.

**Comment by olafgrabienski** [comment]
> Current implementation:
> 1. hover over the admin bar "this page" item 
> 2. hover over the layout name item 
> 3. click "configure" -> 1 less action = UX+

@klonos The "Configure" link goes to "Configure Layout" where you choose the template etc. The actual steps to the "Manage blocks" page are even less:
1. hover over the admin bar "this page" item 
2. click the layout name item

**Comment by indigoxela** [comment]
What's the actual status of this issue and PR? We have a "works for me" label, lots of comments, general support for the approach, but some concerns (or one concern?).

Is there something that needs to get addressed, or are we just bike-sheding? :wink:

**Comment by BWPanda** [comment]
I think the concerns raised can be discussed in a followup issue once (if?) this gets in. Otherwise just needs code review I believe.

**Comment by klonos** [comment]
> @klonos The "Configure" link points to "Configure Layout" where you choose the template etc. The actual steps to the "Manage blocks" page are even less:

I stand corrected 🙇🏼

**Comment by docwilmot** [comment]
Once this is in, I'm thinking of a contrib "Help" module that matches page url to a docs chapter or section and puts a link in that menu.

**Comment by BWPanda** [comment]
PR updated and back to needs review/testing.

**Comment by olafgrabienski** [comment]
@BWPanda Had a quick look at the sandbox: on a phone or other narrow screen I can't reach the links to the Layout and the Theme: Under "More tasks" I see "This page", but when I tap the item, nothing happens. (Reproducible with Dev tools in Browser.)

**Comment by BWPanda** [comment]
@olafgrabienski Thanks for testing. As mentioned [above](https://github.com/backdrop/backdrop-issues/issues/2626#issuecomment-998456495), the issue of the drop-down menu going off the page is an existing issue with the Admin Bar (https://github.com/backdrop/backdrop-issues/issues/4549) and as such I don't think it should hold up this issue.

**Comment by olafgrabienski** [comment]
@BWPanda Sorry, I didn't realize it was the same issue as 4549. Behaves a bit different in the "More tasks" dropdown and with a not clickable parent item, but now I see it's the same.

**Comment by stpaultim** [comment]
I am so anxious to see this get into core! I tested today and I think it's working as expected. I'm going to mark WFM.

**Comment by indigoxela** [comment]
Cool, we're almost done. One problem: tests on php 5.3 fail. I'm aware that we're dropping support for that old version, but that's not merged yet, so the tests should still pass.

Temporarily back to "needs work".

**Comment by BWPanda** [comment]
All tests passing now, thanks to @indigoxela's suggestion 👍🏻

**Comment by indigoxela** [comment]
Great work, RTBC!

Do we already have an issue for the hidden admin menu submenus on narrow screens? If not, IMO we should open one.

**Comment by BWPanda** [comment]
@indigoxela We do: https://github.com/backdrop/backdrop-issues/issues/4549 (was also linked above).

**Comment by BWPanda** [comment]
FYI, the PR in https://github.com/backdrop/backdrop-issues/issues/4549 also fixes the problem here too:

![Peek 2022-03-25 08-27](https://user-images.githubusercontent.com/2385329/160013372-1a61613a-29fe-41cb-8e61-c9c5b0ffd9b6.gif)

**Comment by olafgrabienski** [comment]
> the PR in https://github.com/backdrop/backdrop-issues/issues/4549 also fixes the problem here too

That's great news, thank you very much!

**Comment by laryn** [comment]
Wow, I filed this almost exactly 5 years ago and that was a lot of comments, PRs, feedback and review. Great job, team! Thanks for the PRs so we could thoroughly test and review two different approaches, @BWPanda and @bugfolder. (As noted, the contextual link approach is still available for those who prefer it with a [contrib module](https://github.com/backdrop-contrib/layout_contextual_links) and the option to turn off the admin menu link).

Thanks for the code review @indigoxela as well as testing and feedback from @klonos, @stpaultim, @jenlampton, @laryn, @docwilmot, @olafgrabienski, @philsward, and @Graham-72.

Pulling out this comment from @docwilmot as a potential follow-up project:

> Once this is in, I'm thinking of a contrib "Help" module that matches page url to a docs chapter or section and puts a link in that menu.
